src/nemos/basis.py

```diff
@@ -30,36 +32,333 @@ def __dir__() -> list[str]:
+    """Check input before calling basis.
+
+    This decorator allows to raise an exception that is more readable
+    when the wrong number of input is provided to __call__.
     """
-    Abstract class for basis functions.
+
+    def wrapper(self: Basis, *xi: ArrayLike, **kwargs) -> NDArray:
+        xi = self._check_transform_input(*xi)
+        return func(self, *xi, **kwargs)  # Call the basis
+
+    return wrapper
+
+
+def check_one_dimensional(func: Callable) -> Callable:
+    def wrapper(self: Basis, *xi: ArrayLike, **kwargs):
+        if any(x.ndim != 1 for x in xi):
+            raise ValueError("Input sample must be one dimensional!")
+        return func(self, *xi, **kwargs)
+
+    return wrapper
+
+
+def min_max_rescale_samples(sample_pts: NDArray) -> NDArray:
+    if np.any(sample_pts < 0) or np.any(sample_pts > 1):
+        sample_pts -= np.min(sample_pts)
+        sample_pts /= np.max(sample_pts)
+        warnings.warn(
+            "Rescaling sample points for RaisedCosine basis to [0,1]!", UserWarning
+        )
+    return sample_pts
+
+
+class TransformerBasis:
+    """Basis as `scikit-learn` transformers.
+
+    This class abstracts the underlying basis function details, offering methods
+    similar to scikit-learn's transformers but specifically designed for basis
+    transformations. It supports fitting to data (calculating any necessary parameters
+    of the basis functions), transforming data (applying the basis functions to
+    data), and both fitting and transforming in one step.
 
     Parameters
     ----------
-    n_basis_funcs :
-        Number of basis functions.
+    basis : Basis
+        An instance of a Basis class or any subclass thereof.
+    """
 
-    Attributes
+    def __init__(self, basis: Basis):
+        self._basis = basis
+
+    def fit(self, X: NDArray, y=None):
+        """
+        Compute the convolutional kernels.
+
+        If any of the 1D basis in self._basis is in "conv" mode, it computes the convolutional kernels.
+
+        Parameters
+        ----------
+        X :
+           The data to fit the basis functions to, shape (num_samples, num_input).
+        y : ignored
+           Not used, present for API consistency by convention.
+
+        Returns
+        -------
+        self :
+            The transformer object.
+        """
+        self._basis._set_kernel(*X.T)
+        return self
+
+    def transform(self, X: NDArray, y=None) -> NDArray:
+        """
+        Transform the data using the fitted basis functions.
+
+        Parameters
+        ----------
+        X : array-like
+            The data to transform using the basis functions, shape (num_samples, num_input).
+        y : ignored
+            Not used, present for API consistency by convention.
+
+        Returns
+        -------
+        :
+            The data transformed by the basis functions.
+        """
+        return self._basis._compute_features(*X.T)
```

Comment:
I figured decorating would not work because the convolution requires pynapple to be passed to the function, so it can take care of multiple epochs, If I decorate transformations, then the convolution would receive an array with the data and do not perform the correct operation.

The original `Basis` object support `pynapple` already, so what I ended up doing was creating an "unpacking" method that does return `(X[:, k] for k in X.shape[1])`, this will return pynapple Tsds in a generator and that is what will be passed to `Basis`